src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_internal.clj

```diff
@@ -86,6 +86,29 @@
    pool :- jruby-schemas/pool-queue-type]
   (.borrowItemWithTimeout pool timeout TimeUnit/MILLISECONDS))
 
+(schema/defn insert-shutdown-poison-pill
+  [pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (ShutdownPoisonPill. pool)))
+
+(schema/defn insert-poison-pill
+  [error :- Throwable
+   pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (PoisonPill. error)))
+
+(schema/defn fill-pool-with-poison-pills!
+  "Fill an empty pool with poison pills. Should always be run on the
+  modify-instance-agent. insert-fn should be a fn that inserts a poison
+  pill of some kind into the pool"
+  [pool-state :- jruby-schemas/PoolState
+   insert-fn :- IFn]
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)]
+    (when (not (= pool-size (.remainingCapacity pool)))
```

Comment:
> In the lock call where we await the allRegisteredInQueue Condition if the pool isn't full, short-circuit the await either if there the pill is non-null or if the pool has been refilled.

Assuming we go with something like the above... since the insertion of a pill should mean that the pool is effectively dead, I wonder if it might be good to have the `lock` call throw an `Exception` - internally releasing the pool lock in the process - if the `lock` is short-circuited by the insertion of a pill into the pool.  The `lock` method already can throw an `InterruptedException` if the thread on which the `lock` call is made is interrupted - which should only happen during a prolonged shutdown.

I think it might be bad if the `lock` call were to succeed, for example, if called by file sync after a pill has been inserted.  For example, file sync might otherwise be granted a lock on a pool that was only partially initialized and update code on disk while a "good" borrowed instance is in the middle of compiling a catalog.  If an exception is thrown from `lock` early (because the pool is borked and we need to shut down) file sync could then presumably bail out on applying a code update as well.

@KevinCorcoran - would definitely like your sanity check on this bit.  If this isn't making any sense, maybe we can get together for a few minutes after the holidays and go over this in person.